engine/src/main/java/org/apache/apex/engine/EmbeddedAppLauncherImpl.java

```diff
@@ -54,6 +58,56 @@ public DAG cloneDAG() throws Exception
   }
 
   @Override
+  public EmbeddedAppHandleImpl launchApp(StreamingApplication application, Configuration configuration, Attribute.AttributeMap
+      launchParameters) throws LauncherException
+  {
+    try {
+      prepareDAG(application, configuration);
+    } catch (Exception e) {
+      throw new LauncherException(e);
+    }
+    LocalMode.Controller lc = getController();
+    boolean launched = false;
+    if (launchParameters != null) {
+      if (launchParameters.getValue(SERIALIZE_DAG)) {
+        // Check if DAG can be serialized
+        try {
+          cloneDAG();
+        } catch (Exception e) {
+          throw new LauncherException(e);
+        }
+      }
+      if (launchParameters.getValue(HEARTBEAT_MONITORING)) {
+        lc.setHeartbeatMonitoringEnabled(true);
+      }
+      if (launchParameters.getValue(RUN_ASYNC)) {
+        lc.runAsync();
+        launched = true;
+      } else {
+        Long runMillis = launchParameters.getValue(RUN_MILLIS);
+        if (runMillis != null) {
+          lc.run(runMillis);
+          launched = true;
+        }
+      }
+    }
+    if (!launched) {
+      lc.run();
+    }
+    return new EmbeddedAppHandleImpl(lc);
+  }
+
+  @Override
+  public void shutdownApp(EmbeddedAppHandleImpl app, ShutdownMode shutdownMode)
+  {
+    if (shutdownMode != ShutdownMode.KILL) {
+      app.controller.shutdown();
```

Comment:
Recently we added waiting for the local mode threads to finish running isn't it. So wouldn't it be considered an AWAIT_TERMINATION